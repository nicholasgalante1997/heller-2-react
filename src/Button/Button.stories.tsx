import React from 'react';
import { type Meta, type StoryObj } from '@storybook/react';
import { colorBaseBlackAlpha, colorBaseWhiteAlpha } from 'heller-2-lite';
import { Button } from './Button';

const meta: Meta<typeof Button> = {
  title: 'react/components/Button',
  component: Button
};

export default meta;

type ButtonStoryType = StoryObj<typeof Button>;

const containerStyle = (mode: string) =>
  ({
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
    gap: '16px',
    boxSizing: 'border-box',
    margin: 0,
    padding: '16px',
    background: mode === 'light' ? colorBaseWhiteAlpha : colorBaseBlackAlpha,
    height: '500px',
    width: '500px'
  }) as React.CSSProperties;

export const PrimaryLight: ButtonStoryType = {
  decorators: [
    (Story) => (
      <div className="heller2-theme" data-theme-mode="light" style={containerStyle('light')}>
        <Story />
      </div>
    )
  ],
  render: (args) => <Button {...args}>Click Me</Button>
};

export const SecondaryLight: ButtonStoryType = {
  decorators: [
    (Story) => (
      <div className="heller2-theme" data-theme-mode="light" style={containerStyle('light')}>
        <Story />
      </div>
    )
  ],
  render: (args) => (
    <Button v="secondary" {...args}>
      Click Me
    </Button>
  )
};

export const TertiaryLight: ButtonStoryType = {
  decorators: [
    (Story) => (
      <div className="heller2-theme" data-theme-mode="light" style={containerStyle('light')}>
        <Story />
      </div>
    )
  ],
  render: (args) => (
    <Button v="tertiary" {...args}>
      Click Me
    </Button>
  )
};

export const NoirLight: ButtonStoryType = {
  decorators: [
    (Story) => (
      <div className="heller2-theme" data-theme-mode="light" style={containerStyle('light')}>
        <Story />
      </div>
    )
  ],
  render: (args) => (
    <Button v="noir" {...args}>
      Click Me
    </Button>
  )
};

export const PrimaryDark: ButtonStoryType = {
  decorators: [
    (Story) => (
      <div className="heller2-theme" data-theme-mode="dark" style={containerStyle('dark')}>
        <Story />
      </div>
    )
  ],
  render: (args) => <Button {...args}>Click Me</Button>
};

export const SecondaryDark: ButtonStoryType = {
  decorators: [
    (Story) => (
      <div className="heller2-theme" data-theme-mode="dark" style={containerStyle('dark')}>
        <Story />
      </div>
    )
  ],
  render: (args) => (
    <Button v="secondary" {...args}>
      Click Me
    </Button>
  )
};

export const TertiaryDark: ButtonStoryType = {
  decorators: [
    (Story) => (
      <div className="heller2-theme" data-theme-mode="dark" style={containerStyle('dark')}>
        <Story />
      </div>
    )
  ],
  render: (args) => (
    <Button v="tertiary" {...args}>
      Click Me
    </Button>
  )
};

export const NoirDark: ButtonStoryType = {
  decorators: [
    (Story) => (
      <div className="heller2-theme" data-theme-mode="dark" style={containerStyle('dark')}>
        <Story />
      </div>
    )
  ],
  render: (args) => (
    <Button v="noir" {...args}>
      Click Me
    </Button>
  )
};

export const Hover_RaiseElevation: ButtonStoryType = {
  decorators: [
    (Story) => (
      <div className="heller2-theme" data-theme-mode="light" style={containerStyle('light')}>
        <Story />
      </div>
    )
  ],
  render: (args) => (
    <Button v="primary" size="medium" hover={{ animationType: 'raise-elevation' }} {...args}>
      Click Me
    </Button>
  )
};

export const Hover_TransitionBG: ButtonStoryType = {
  decorators: [
    (Story) => (
      <div className="heller2-theme" data-theme-mode="light" style={containerStyle('light')}>
        <Story />
      </div>
    )
  ],
  render: (args) => (
    <Button v="primary" size="medium" hover={{ animationType: 'background-transition' }} {...args}>
      Click Me
    </Button>
  )
};
