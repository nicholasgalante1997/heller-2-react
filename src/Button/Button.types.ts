import React from 'react';
import { type HasChildNodes } from '@/types/HasChildren';

interface InternalButtonProps {
  v?: 'primary' | 'secondary' | 'tertiary' | 'noir';
  size?: 'small' | 'medium' | 'large';
  hover?: {
    animationType: 'raise-elevation' | 'background-transition';
  }
}

export type ButtonProps = Omit<React.HTMLProps<HTMLButtonElement>, 'size' | 'type'> & HasChildNodes & InternalButtonProps;