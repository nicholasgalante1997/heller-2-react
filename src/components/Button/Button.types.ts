import type React from 'react';
import { type HasChildNodes } from '@/types/HasChildren';

export type ButtonAnimationOnHover = 'raise-elevation' | 'background-transition' | 'scale-content';
interface InternalButtonProps {
  v?: 'primary' | 'secondary' | 'tertiary' | 'noir';
  size?: 'small' | 'medium' | 'large';
  rounded?: boolean;
  hover?: {
    animationType: ButtonAnimationOnHover;
  };
}

export type ButtonProps = Omit<React.HTMLProps<HTMLButtonElement>, 'size' | 'type'> &
  HasChildNodes &
  InternalButtonProps;
