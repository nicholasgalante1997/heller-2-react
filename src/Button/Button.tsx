import React from 'react';
import classNames from 'classnames';

import { ButtonProps } from './Button.types';

const buttonVariantClassNameMap = new Map<ButtonProps['v'], string>([
  ['primary', 'h-button-primary'],
  ['secondary', 'h-button-secondary'],
  ['tertiary', 'h-button-tertiary'],
  ['noir', 'h-button-noir']
]);

const buttonSizeClassMap = new Map<ButtonProps['size'], string>([
  ['small', 'h-button-size-small'],
  ['medium', 'h-button-size-medium'],
  ['large', 'h-button-size-large']
]);

function ButtonComponent({
  children,
  hover,
  v = 'primary',
  size = 'medium',
  className: classNameProp = '',
  ...rest
}: ButtonProps) {
  const buttonDefaultClassName = 'h-button';
  const buttonVClassName = React.useMemo(() => buttonVariantClassNameMap.get(v), [v]);
  const buttonSizeClassName = React.useMemo(() => buttonSizeClassMap.get(size), [size]);
  const buttonHoverClassName = React.useMemo(() => {
    if (hover) {
      if (hover.animationType === 'raise-elevation') {
        return 'hover-raise-elevation-100';
      } else {
        return `h-button-hover-${v}`;
      }
    } else {
      return '';
    }
  }, [hover]);
  const className = React.useMemo(
    () =>
      classNames(
        buttonDefaultClassName,
        buttonSizeClassName,
        buttonVClassName,
        buttonHoverClassName,
        classNameProp
      ),
    [buttonVClassName, buttonSizeClassName, classNameProp, buttonHoverClassName]
  );
  return (
    <button className={className} {...rest}>
      {children}
    </button>
  );
}

export const Button = React.memo(ButtonComponent);
