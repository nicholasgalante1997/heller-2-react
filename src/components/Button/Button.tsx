import React from 'react';
import classNames from 'classnames';

import { type ButtonProps, type ButtonAnimationOnHover } from './Button.types';

const buttonVariantClassNameMap = new Map<ButtonProps['v'], string>([
  ['primary', 'h-button-primary'],
  ['secondary', 'h-button-secondary'],
  ['tertiary', 'h-button-tertiary'],
  ['noir', 'h-button-noir']
]);

const buttonSizeClassMap = new Map<ButtonProps['size'], string>([
  ['small', 'h-button-size-small'],
  ['medium', 'h-button-size-medium'],
  ['large', 'h-button-size-large']
]);

const buttonAnimationOnHoverClassMap = new Map<ButtonAnimationOnHover, string>([
  ['background-transition', 'h-button-hover'],
  ['raise-elevation', 'hover-raise-elevation-100'],
  ['scale-content', 'hover-scale-200']
]);

function getHoverAnimation(type: ButtonAnimationOnHover, variant: ButtonProps['v'] = 'primary') {
  if (type === 'background-transition') {
    return buttonAnimationOnHoverClassMap.get(type) + `-${variant}`;
  } else {
    return buttonAnimationOnHoverClassMap.get(type);
  }
}

function ButtonComponent({
  children,
  hover,
  v = 'primary',
  size = 'medium',
  className: classNameProp = '',
  ...rest
}: ButtonProps) {
  const buttonDefaultClassName = 'h-button';
  const buttonVClassName = React.useMemo(() => buttonVariantClassNameMap.get(v), [v]);
  const buttonSizeClassName = React.useMemo(() => buttonSizeClassMap.get(size), [size]);
  const buttonHoverClassName = React.useMemo(() => hover && getHoverAnimation(hover.animationType), [hover]);
  const className = React.useMemo(
    () =>
      classNames(
        buttonDefaultClassName,
        buttonSizeClassName,
        buttonVClassName,
        buttonHoverClassName,
        classNameProp
      ),
    [buttonVClassName, buttonSizeClassName, classNameProp, buttonHoverClassName]
  );
  return (
    <button className={className} {...rest}>
      {children}
    </button>
  );
}

export const Button = React.memo(ButtonComponent);
