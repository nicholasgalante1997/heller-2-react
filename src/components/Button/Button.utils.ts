import { type ButtonAnimationOnHover, type ButtonProps } from './Button.types';

const buttonVariantClassNameMap = new Map<ButtonProps['v'], string>([
  ['primary', 'h-button-primary'],
  ['secondary', 'h-button-secondary'],
  ['tertiary', 'h-button-tertiary'],
  ['noir', 'h-button-noir']
]);

const buttonSizeClassMap = new Map<ButtonProps['size'], string>([
  ['small', 'h-button-size-small'],
  ['medium', 'h-button-size-medium'],
  ['large', 'h-button-size-large']
]);

const buttonAnimationOnHoverClassMap = new Map<ButtonAnimationOnHover, string>([
  ['background-transition', 'h-button-hover'],
  ['raise-elevation', 'hover-raise-elevation-100'],
  ['scale-content', 'hover-scale-200']
]);

function getHoverAnimation(type: ButtonAnimationOnHover, variant: ButtonProps['v'] = 'primary') {
  if (type === 'background-transition') {
    return buttonAnimationOnHoverClassMap.get(type) + `-${variant}`;
  } else {
    return buttonAnimationOnHoverClassMap.get(type);
  }
}

export { buttonAnimationOnHoverClassMap, buttonSizeClassMap, buttonVariantClassNameMap, getHoverAnimation };
