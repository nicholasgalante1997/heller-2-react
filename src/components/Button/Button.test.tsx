import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvents from '@testing-library/user-event';
import { Button } from './Button';

describe('Button', () => {
  describe('Events', () => {
    test('onClick', async () => {
      const testId = 'h-button-id';
      const onClick = jest.fn();
      const jsx = (
        <Button data-testid={testId} onClick={onClick}>
          Shmallas
        </Button>
      );
      render(jsx);
      const button = screen.getByTestId(testId);
      expect(button).not.toBeNull();
      await userEvents.click(button);
      expect(onClick).toHaveBeenCalledTimes(1);
    });
  });
  test('Snapshot', () => {
    const jsx = <Button>Shmallas</Button>;
    const { container } = render(jsx);
    expect(container).toMatchSnapshot();
  });
});
