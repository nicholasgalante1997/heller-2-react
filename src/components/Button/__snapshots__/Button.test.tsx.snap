// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Snapshot 1`] = `
<div>
  <button
    class="h-button h-button-size-medium h-button-primary"
  >
    Shmallas
  </button>
</div>
`;
